src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -249,7 +247,10 @@ public T Dequeue()
 
             T removed = _array[_head];
             _array[_head] = default(T);
-            _head = (_head + 1) % _array.Length;
+            if (++_head == _array.Length)
```

Comment:
@sharwell 
I think the classes not intended to be used in multithreaded manner don't guaranty anything. An excepton instead of returning a wrong value is not a worst thing that can happen to a  
careless dev :)
It is explicitly stated in "Any instance members are not guaranteed to be thread safe."